portfolio/src/main/java/com/google/sps/servlets/LikeCommentServlet.java

```diff
@@ -0,0 +1,129 @@
+import javax.servlet.http.HttpServletResponse;
+
+
+@WebServlet("/toggle-like")
+public class LikeCommentServlet extends HttpServlet {
+
+  private DatastoreService datastore;
+
+  @Override
+  public void init() {
+    datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String commentKeyString = request.getParameter("key");
+    Entity comment = null;
+    try {
+      comment = datastore.get(KeyFactory.stringToKey(commentKeyString));
+    } catch (EntityNotFoundException e) {
+      e.printStackTrace();
+      response.sendRedirect("/");
+      return;
+    }
+
+    // The first element in this list records whether or not the client has like this comment, and the second records the number of likes this comment has.
+    List<String> commentLikeData = new LinkedList<>();
+
+    // If the comment has at least one like... 
+    if (comment.hasProperty("likedBy")){
+      HashSet<String> likedByIpSet = new HashSet<String>((List<String>) comment.getProperty("likedBy"));
+
+      // ...and the client has liked it, record that the client has liked it by adding "true".
+      if (likedByIpSet.contains(request.getRemoteAddr())) {
+        commentLikeData.add("true");
+
+      // ...and the client has not liked it, record that the client has not liked it by adding "false".
+      } else {
+        commentLikeData.add("false");
+      }
+
+      // Then, record the number of likes that the comment has.
+      commentLikeData.add(Integer.toString(likedByIpSet.size()));
+
+    // If the comment has no likes, record "false" since the client hasn't liked it and "0" since it has no likes.
+    } else {
+      commentLikeData.add("false");
+      commentLikeData.add("0");
+    }
+
+    response.setContentType("application/json;");
+    response.getWriter().println(convertToJsonUsingGson(commentLikeData));  
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String commentKeyString = request.getParameter("key");
+    Entity comment = null;
+
+    // If it exists, get the comment from the datastore by its key.
+    try {
+      comment = datastore.get(KeyFactory.stringToKey(commentKeyString));
+    } catch (EntityNotFoundException e) {
+      e.printStackTrace();
+      response.sendRedirect("/");
+      return;
+    }
+
+    String clientIp = request.getRemoteAddr();
+    Set<String> likedByIpSet;
+
+    // If the comment has already been liked at least once, add or remove this client's IP to the set of IPs that have liked it.
+    if (comment.hasProperty("likedBy")) {
+      likedByIpSet = new HashSet<String>((List<String>) comment.getProperty("likedBy"));
+
+      // If the client's IP address is already in the set, then the comment is already liked, and needs to be unliked. 
+      if (likedByIpSet.contains(clientIp)) {
```

Comment:
As discussed, I think ip is not best way to track likes and it would be better to attribute to authenticated users. However, I think this is sufficient for demonstrating the idea/learning and there isn't any need to change it.